"cinema package added"

pChupakhin opened the pull request:
Cinema project created

pChupakhin committed:
"cinema package added"

yvarshavskyigd commented:
Java is object-oriented programming language, please use objects instead of static

pChupakhin committed:
CHANGES FROM THE LAST PULL REQUEST
 1. Cinema, MainCinema, Menu classes created  to avoid non-conventional static
 2. All unnecessary static removed
 3. All classes packed in com.driddynamics.cinema
 4. .gitignore updated

pChupakhin committed:
CHANGES FROM THE LAST PULL REQUEST
 1. Cinema, MainCinema, Menu classes created  to avoid non-conventional static
 2. All unnecessary static removed
 3. All classes packed in com.driddynamics.cinema
 4. cinema package removed
 5. .gitignore updated

pChupakhin commented:
CHANGES FROM THE LAST PULL REQUEST
 1. Cinema, MainCinema, Menu classes created  to avoid non-conventional static
 2. All unnecessary static removed
 3. All classes packed in com.driddynamics.cinema
 4. cinema package removed
 5. .gitignore updated

yvarshavskyigd commented on Cinema Room Manager.iml:
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Do you need this file on GitHub? if not - remove it

pChupakhin replied:
No, there is no need in it on GitHub. File already removed

yvarshavskyigd commented on src/com/griddynamics/cinema/MainCinema.java:
```diff
@@ -0,0 +1,7 @@
+
+public class MainCinema {
+        public static void main(String[] args) {
+        new Menu().cinemaMenu();
```
What is Menu?

pChupakhin replied:
Menu is class, implemented to give user the ability to change state or get information about current state of Cinema class object, by entering number of corresponding menu item in command line.

yvarshavskyigd replied:
call it "cli", would be easier to understand
https://en.wikipedia.org/wiki/Command-line_interface

yvarshavskyigd commented on src/com/griddynamics/cinema/Menu.java:
```diff
@@ -0,0 +1,71 @@
+	
+	private static final Cinema cinema;
+	
+	static {
```
Still using static

yvarshavskyigd commented:
Do you need files `.class`?

yvarshavskyigd commented:
Please reformat a code, hard to read it

pChupakhin committed:
Cinema Room Manager. iml removed
All static removed
Menu class renamed to CinemaMenu
Object of CinemaMenu class declarated
Operations, that change values of Cinema class fields encapsulated in particular methods

pChupakhin committed:
Cinema Room Manager. iml removed
All static removed
Menu class renamed to CinemaMenu
Object of CinemaMenu class declarated
Operations, that change values of Cinema class fields encapsulated in particular methods

pChupakhin committed:
CInema Room Manager.iml removed

pChupakhin committed:
redundant directory removed

pChupakhin commented:
> Do you need files `.class`?



> Do you need files `.class`?



> Do you need files `.class`?


> Do you need files `.class`?

No. Files already removed

pChupakhin commented:
> Do you need files `.class`?

No. Files already removed

yvarshavskyigd commented:
In regard to Reformat the code
https://www.jetbrains.com/help/idea/reformat-and-rearrange-code.html

yvarshavskyigd commented on src/com/griddynamics/cinema/Cinema.java:
```diff
@@ -0,0 +1,82 @@
+import java.util.stream.Stream;
+
+public class Cinema {
+	
```
Please take a look at Naming Conventions
https://www.oracle.com/java/technologies/javase/codeconventions-namingconventions.html

yvarshavskyigd commented on src/com/griddynamics/cinema/Cinema.java:
```diff
@@ -0,0 +1,82 @@
+
+public class Cinema {
+	
+    final char[][] HALL;
```
Think about refactor this `hall` variable, create object for it.  Your business logic depends on representation, it should be opposite (for example, you look at a char - is seat booked or not), which is not good.
My suggestion is to create class Seat, which will contain information about is it reserved or not. And create a HallRow, which will contain list of Seats and row number. You can calculate there reserved seat by each row.

yvarshavskyigd commented on src/com/griddynamics/cinema/Cinema.java:
```diff
@@ -0,0 +1,82 @@
+    void showSeats() {
+
+        System.out.println("\nCinema:\n"
+                + "  1 2 3 4 5 6 7 8 9".substring(0, SEATS_IN_EACH_ROW * 2 + 1));
```
what happens if seats in each row will be 10?

pChupakhin replied:
In description of one of projects stages was declared condition that number of rows and number of seats would be less than 10.

pChupakhin replied:
Will use the "if" statement to filter values greater than 9

yvarshavskyigd replied:
I think it is not the best way how to generate string of numbers, how you can change it?

pChupakhin replied:
Can I use this?
System.out.print(IntStream.generate(new AtomicInteger()::getAndIncrement)
				.limit(100/**/).boxed().map(String::valueOf).collect(Collectors.joining(" ")));

pChupakhin replied:
Can I use this:
System.out.print(
IntStream.generate(new AtomicInteger()::getAndIncrement)
.limit(/*random amount of rows*/)
.boxed()
.map(String::valueOf)
.collect(Collectors.joining(" "))
);
?

pChupakhin replied:
Can I use this:

		System.out.print(
				IntStream.generate(new AtomicInteger()::getAndIncrement)
				.limit(/*random amount of rows*/)
				.boxed()
				.map(String::valueOf)
				.collect(Collectors.joining(" ")));

?

pChupakhin replied:
Can I use this:

	System.out.print(
			IntStream.generate(new AtomicInteger()::getAndIncrement)
			.limit(/*random amount of seats*/)
			.boxed()
			.map(String::valueOf)
			.collect(Collectors.joining(" ")));
?

pChupakhin replied:
By using Stream API, I think. For example:

		 System.out.println(Stream.iterate(1, i -> i + 1)
				 .limit(/*random amount of seats*/)
				 .map(String::vlaueOf)
				 .collect(Collectors.joining(" ", "   ", "")));

pChupakhin replied:
By using Stream API, I think. For example:

    	System.out.println("\nCinema\n"
    		+ Stream.iterate(1, i -> i + 1)
    			.limit(/*random amount of seats*/1)
    			.map(String::valueOf)
    			.collect(Collectors.joining(" ", "   ", "")));

pChupakhin replied:
By using Stream API, I think. For example:

	System.out.println("\nCinema\n"
		+ Stream.iterate(1, i -> i + 1)
			.limit(/*random amount of seats*/)
			.map(String::valueOf)
			.collect(Collectors.joining(" ", "   ", "")));

yvarshavskyigd reviewed (approved):
.

pChupakhin committed:
Delete .gitignore

.gitignore removed

pChupakhin committed:
- Cinema class updated
 - CinemaMenu class removed
 - CinemaCli class added
 - MainCinema class removed
 - Main class added
 - Hall class added
 - Row class added
 - Seat class added

pChupakhin committed:
Merge remote-tracking branch 'origin/dev' into dev

pChupakhin committed:
CinemaCli class updated

pChupakhin committed:
CinemaCli class updated

andrew4herman commented on src/com/griddynamics/cinema/Seat.java:
```diff
@@ -0,0 +1,19 @@
+public class Seat {
+
+    private boolean isReserved = false;
+    
```
How about rename this field to bare `reserved` instead of `isReserved`?

`Is` is a verb and verbs are inappropriate to represent an Object's state. Use an adjective instead, and suddenly your getter / setter names will make more sense (you have already named them correctly):

`isReserved()`
`reserve()`

[resource](https://stackoverflow.com/questions/5322648/for-a-boolean-field-what-is-the-naming-convention-for-its-getter-setter)

pChupakhin replied:
got it!

andrew4herman commented on src/com/griddynamics/cinema/Row.java:
```diff
@@ -0,0 +1,24 @@
+
+    public List<Seat> getSeatsInEachRow() {
+        return seatsInEachRow;
+    }
```
What about this field `seatsInEachRow`?
Does it really contain seats in each row? Or it knows only amount of seats from its row?
I suggest you rename it to `seatsInRow` and to `getSeatsInRow()` accordingly.

yvarshavskyigd replied:
Just `seats`. It should be enough 🙂

pChupakhin replied:
it contains seats in each row, but your suggestion accepted. names will be changed

andrew4herman commented on src/com/griddynamics/cinema/Main.java:
```diff
@@ -0,0 +1,41 @@
+        
+        Cinema cinema;
+        CinemaCli cli;
+    
```
Looks like it should be on lines 37 and 38

pChupakhin replied:
got it!

pChupakhin committed:
code style changed

pChupakhin commented:
Changes From The Last Pull Request:
1. All .class files removed
2. Code style changed to more readable format
3. Cinema Room Manager.iml file removed
4. Menu class renamed to CinemaCli
5. All unnecessary static removed
6. All packages, classes, methods, variables, constants renamed according to naming convention
7. .gitignore file deleted
8. Business logic refactored according to OOP

andrew4herman commented on src/com/griddynamics/cinema/CinemaCli.java:
```diff
@@ -0,0 +1,134 @@
+
+import java.util.Scanner;
+
+public class CinemaCli {
+    
+    private static final int ONE_HUNDRED_PERCENT = 100;
+    
```
Do you really need this field? I guess It is obvious enough when you divide by 100 to receive _percent_.

yvarshavskyigd replied:
It calls magic numbers 
https://refactoring.guru/replace-magic-number-with-symbolic-constant
This discussion is interesting, please take a look at it.
https://codereview.stackexchange.com/questions/44047/100-is-a-magic-number

I think that we can avoid this constant, because looking at "100" I understand, that it is one hundred. But sometimes you can have a static program analysis tool on the project, which will fail because of "magic number"

andrew4herman replied:
Oh, thank you, I will definitely take a look at that

andrew4herman commented on src/com/griddynamics/cinema/CinemaCli.java:
```diff
@@ -0,0 +1,134 @@
+    private static final int LARGE_CINEMA_HALL_MINIMUM_SEATS_AMOUNT = 60;
+    private static final int TICKET_FULL_PRICE = 10;
+    private static final int TICKET_DISCOUNT_PRICE = 8;
+    
+    private final Scanner scanner = new Scanner(System.in);
+    
```
Can you move Scanner initialization to the constructor body? I think that would be much clearer

pChupakhin replied:
I can, but it is not necessary

andrew4herman commented on src/com/griddynamics/cinema/CinemaCli.java:
```diff
@@ -0,0 +1,134 @@
+    
+    private static final int ONE_HUNDRED_PERCENT = 100;
+    
+    private static final int LARGE_CINEMA_HALL_MINIMUM_SEATS_AMOUNT = 60;
```
You are using this field only in constructor to set another boolean field. I think that field is unnecessary and can be avoided.

andrew4herman replied:
Second one

yvarshavskyigd replied:
Here is magic numbers, please take a look at my [comment](https://github.com/pChupakhin/Cinema_Room_Manager/pull/5#discussion_r772704598) above. I suggest to keep it as it is 🙂

andrew4herman commented on src/com/griddynamics/cinema/CinemaCli.java:
```diff
@@ -0,0 +1,134 @@
+        
+        isLargeCinemaHall = totalSeatsAmount >= LARGE_CINEMA_HALL_MINIMUM_SEATS_AMOUNT;
+    }
+    
+    public Cinema getCinema() {
+        return cinema;
+    }
+    
```
You invoke this method only in the same class where this field is declared. I guess you can remove it and get `cinema` object directly from the field. Or at least make this method to be private.

pChupakhin replied:
Thank you, but I will leave it as is.

andrew4herman commented on src/com/griddynamics/cinema/CinemaCli.java:
```diff
@@ -0,0 +1,134 @@
+            }
+        } while (userOption != 0);
+    }
+    
+    private void ticketPurchasingCLI() throws RuntimeException {
+        int row;
```
RuntimeException and its subclasses are unchecked exceptions. Unchecked exceptions do not need to be declared in a method or constructor's throws clause if they can be thrown by the execution of the method or constructor and propagate outside the method or constructor boundary.

pChupakhin replied:
Got it!

andrew4herman commented on src/com/griddynamics/cinema/CinemaCli.java:
```diff
@@ -0,0 +1,134 @@
+        return (double)getPurchasedTicketsAmount() / totalSeatsAmount * ONE_HUNDRED_PERCENT;
+    }
+    
+    private int getCurrentIncome() {
+        int purchasedDiscountTicketsAmount = !isLargeCinemaHall ? 0 : (int)getCinema().getHall()
+                                                                                      .getRows()
+                                                                                      .stream()
+                                                                                      .skip(amountOfRowsWithFullPriceSeats)
+                                                                                      .mapToLong(row -> row.getSeatsInEachRow()
+                                                                                                           .stream()
+                                                                                                           .filter(Seat::isReserved)
+                                                                                                           .count())
+                                                                                      .sum();
```
That style is great until we don't use Command + Alt + L.
Do we really need to create stream chain methods in that way?

pChupakhin replied:
Actually we don't. I will change it to more readable format. Thank you!

andrew4herman commented on src/com/griddynamics/cinema/Hall.java:
```diff
@@ -0,0 +1,29 @@
+    }
+    
+    @Override
+    public String toString() {
+        String seatsPositionsLine = Stream.iterate(1, i -> i + 1)
+                                          .limit(getRows().get(0).getSeatsInEachRow().size())
+                                          .map(String::valueOf)
+                                          .collect(Collectors.joining(" ", "  ", "\n"));
+        
```
The same. IDEA moved that methods to the left when it format content.

pChupakhin replied:
Got it!

andrew4herman reviewed (changes requested):
Hey! I have made a small review of your code. Note that I am a such intern as you and also learning how to write code properly. So if you don't agree with any changes feel free to say that :)

yvarshavskyigd commented on src/com/griddynamics/cinema/CinemaCli.java:
```diff
@@ -0,0 +1,134 @@
+    
+    private final boolean isLargeCinemaHall;
+    
+    private final int amountOfRows;
```
Redundant variable, you have `cinema`, which has that information.
Please think about others fields in this context

pChupakhin replied:
Got it!

yvarshavskyigd commented on src/com/griddynamics/cinema/CinemaCli.java:
```diff
@@ -0,0 +1,134 @@
+        amountOfRowsWithFullPriceSeats = amountOfRows / 2;
+        amountOfRowsWithDiscountPriceSeats = (amountOfRows + 1) / 2;
+        
+        isLargeCinemaHall = totalSeatsAmount >= LARGE_CINEMA_HALL_MINIMUM_SEATS_AMOUNT;
```
For me looks like it should be method

pChupakhin replied:
Got it!

yvarshavskyigd commented on src/com/griddynamics/cinema/CinemaCli.java:
```diff
@@ -0,0 +1,134 @@
+    
+    private final Cinema cinema;
+    
+    public CinemaCli(final Cinema cinema) {
```
About the naming, sorry for misleading a bit, if you have cinema as input for this class, then it should not call `Cli`.
This object operates cinema. Let's call it CinemaOperator, what do you think? Feel free to suggest anything else

pChupakhin replied:
It's OK. I will change the name according to your advice.

yvarshavskyigd commented on src/com/griddynamics/cinema/Main.java:
```diff
@@ -0,0 +1,42 @@
+                if (rows > 9 || seatsInEachRow > 9 || rows < 1 || seatsInEachRow < 1) {
+                    throw new RuntimeException("rows/seats amount must be less than 10 and greater than 0");
+                }
+            } catch (RuntimeException e) {
```
Catching unchecked exception is not a good idea,
looks like you don't need try|catch block at all

pChupakhin replied:
Got it!

pChupakhin committed:
- “isReserved” field name changed to “reserved”;
- All occurrence of “seatsInEachRow” in names of fields/methods changed to “seats”;
- “CinemaCli” class changed to “CinemaOperator”;
- Declaration of objects of Cinema and CinemaOperator classes moved to line, in which they are initialized;
- Constant “ONE_HUNDRED_PERCENT” removed;
- All exception declarations and “try catch” blocks removed;
- AutoFormatter settings changed and code reformatted according to it
- Misplaced fields in body of CinemaOperator class replaced by corresponding methods and moved to Hall class body

pChupakhin commented:
Changes From The Last Pull Request:
1. “isReserved” field name changed to “reserved”;
2. All occurrence of “seatsInEachRow” in names of fields/methods changed to “seats”;
3. “CinemaCli” class changed to “CinemaOperator”;
4. Declaration of objects of Cinema and CinemaOperator classes moved to line, in which they are initialized;
5. Constant “ONE_HUNDRED_PERCENT” removed;
6. All exception declarations and “try catch” blocks removed;
7. AutoFormatter settings changed and code reformatted according to it
8. Misplaced fields in body of CinemaOperator class replaced with corresponding methods and moved to Hall class body

pChupakhin committed:
access modifier of "rowPosition" field changed to private

taras-nishchymenko commented on src/com/griddynamics/cinema/CinemaOperator.java:
```diff
@@ -0,0 +1,119 @@
+        return (int)getCinema().getHall()
+                .getRows()
+                .stream()
+                .mapToLong(row -> row.getSeats().stream().filter(Seat::isReserved).count())
```
Refactor it in a single stream

pChupakhin replied:
Got it

taras-nishchymenko commented on src/com/griddynamics/cinema/HallRow.java:
```diff
@@ -0,0 +1,36 @@
+public class HallRow {
+    
+    private final List<Seat> seats;
+    int rowPosition;
```
Why it’s package-private?

pChupakhin replied:
Didn't notice this mistake. It will be changed to private

pChupakhin committed:
1. all constant variables, fields and arguments marked as "final"
2. business logic of "cinemaOperator" class changed

pChupakhin commented:
Changes From The Last Pull Request:
1. all constant variables, fields and arguments marked as "final" 
2. business logic of "cinemaOperator" class changed
